### PR TITLE
Issue #3 - Create or view shopping list using token 

### LINKED ISSUE #3: 3. As a user, I want to set up a new shopping list so I can start tracking purchased items.
A shopping list consists of a set of items associated with a user’s token. Creating a new list consists of the following sequence:

* Generate a new, unique token
* Save the token to `localStorage`
* Show the user the list view

The following script should be used to generate a suitable token: `src/lib/tokens.js`

AC:

* For users who do not already have a token/list, a button or link exists on the home screen that allows them to create a new list
* Clicking the button/link generates a new token and saves it to `localStorage`
* Once the token has been created and saved, the user is redirected to the "list" view
* For users who do have a token already saved in `localStorage`, they should be shown/redirected to the "list" view

### PR DESCRIPTION
**Pull request by jessicasalbert**
_For an example of how to fill this template out, [see this Pull Request](https://github.com/the-collab-lab/tcl-3-smart-shopping-list/pull/44)._ 

## Description

This PR will allow users to create or view a shopping list. When a user opens the app, it will check if a token is stored in localStorage. If so, the user will be directed to the shopping list view. Otherwise, they will be redirected to the home page. On the home page, a "create new list" button is present. Upon being clicked, this button will create a new, unique token using the tokens.js script and save the token to local storage. The user will then be redirected to the shopping list view. 

To ensure that users can't access "ListView" URL without a token - we added a check inside of ListView component to look for token and redirect to Home if no token is found. We are still considering if there are more effective ways of doing this check. (We tried to make the check in the App component but had issues with using the useHistory hook inside of useEffect). 

## Related Issue

Closes #3 

## Acceptance Criteria

-For users who do not already have a token/list, a button or link exists on the home screen that allows them to create a new list
-Clicking the button/link generates a new token and saves it to localStorage
-Once the token has been created and saved, the user is redirected to the "list" view
-For users who do have a token already saved in localStorage, they should be shown/redirected to the "list" view

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

### Before

Screen Shot 2021-01-12 at 12 47 24 PM

### After
[Local store demo](https://gyazo.com/bd6495136ce1cdf11be8a192a062894a)

## Testing Steps / QA Criteria

- Clone this branch and run "npm install" to update dependencies
- Run "npm start" to start the app
- If no token is present in local storage, app should direct to "/" route and "create a new list" button should be present on screen
- Once "create a new list" button is clicked, app should redirect to "/ListView" route and token should be saved into local storage
- Refresh app with token saved in local storage - should direct right to "/ListView" on reload

### CODE REVIEW

File: src/components/ListView.js [hunk 1]
```diff
@@ -1,8 +1,19 @@
...
+const ListView = () => {
+  const history = useHistory();
+
+  useEffect(() => {
+    if (!localStorage.getItem('token')) {
+      history.push('/');
+    } else {
+      history.push('/ListView');
+    }
+  });
```
Comment (Maeesha-Rahman) [inline src/components/ListView.js]: Great work you two!! :smiley: 

Just one suggestion - I'm assuming you only want this to run upon initial load? If that is the case we could provide an empty array as a second argument to useEffect.

```
  useEffect(() => {
    if (!localStorage.getItem('token')) {
      history.push('/');
    } else {
      history.push('/ListView');
    }
  }, []);
```

**Review by Maeesha-Rahman — APPROVED** [review]

File: src/App.js [hunk 2]
```diff
@@ -55,20 +63,4 @@ function MenuLink({ label, to, activeWhenExact = false }) {
...
-}
-
-function AddItem() {
-  return (
-    <div>
-      <h3>Add Item</h3>
-    </div>
-  );
-}
-
```
Comment (thekidnamedkd) [inline src/App.js:73]: Nice clean up here!

File: src/components/Home.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+  const generateToken = () => {
+    const token = getToken();
+    localStorage.setItem('token', token);
+    history.push('/ListView');
+  };
+
+  return <button onClick={generateToken}>Create a new list</button>;
+};
+
+export default Home;
```
Comment (thekidnamedkd) [inline src/components/Home.js:17]: I found this component really nifty and clean! It works very nicely with our router.

File: src/components/ListView.js (hunk 1 above)
Comment (thekidnamedkd) [inline src/components/ListView.js]: And I love this redirect check in conjunction. 😎

**Review by thekidnamedkd — APPROVED** [review]
Effective work! With the extra collaboration today cross team to anticipate merge changes, I think we're going to be in great shape.

File: src/App.js [hunk 4]
```diff
@@ -7,9 +7,11 @@ import {
   Route,
   Link,
   useRouteMatch,
+  Redirect,
 } from 'react-router-dom';
```
Comment (MelinaMejia95) [inline src/App.js:11]: What do you think about adding a space after this line? So libraries and what is created by you are separated; it's something I use to do to keep code cleaner and more organized :D

File: src/components/ListView.js (hunk 1 above)
Comment (MelinaMejia95) [reply src/components/ListView.js]: Agree with this 😸

Comment (jessicasalbert) [reply src/components/ListView.js]: https://reactjs.org/docs/hooks-faq.html#is-it-safe-to-omit-functions-from-the-list-of-dependencies

I got a warning when I added the empty array, so based on this doc I added 'history' as a dependency in the array and it seemed to clear it up. Does anyone know if this makes a difference?

**Review by skylerwshaw — APPROVED** [review]
Looks good to me. There are definitely a few cleanup opportunities including the one highlighted by @MelinaMejia95 but none are blockers to the acceptance criteria.

It looks like https://github.com/the-collab-lab/tcl-16-smart-shopping-list/pull/19 will build on this to utilize the token we're now setting. It was great to hear about the inter-team collaboration this week. Solid!

wave_clap

**Pull request merged**